File: ToDoModel/Email Utilities/SortItemsToExistingFolder.vb

Imports System.IO
Imports Microsoft.Office.Interop
Imports Microsoft.Office.Interop.Outlook
Imports UtilitiesVB
Imports UtilitiesCS

Public Module SortItemsToExistingFolder
    Public Sub InitializeSortToExisting(Optional InitType As String = "Sort",
        Optional QuickLoad As Boolean = False,
        Optional WholeConversation As Boolean = True,
        Optional strSeed As String = "",
        Optional objItem As Object = Nothing)
        Throw New NotImplementedException
    End Sub

    Public Sub MASTER_SortEmailsToExistingFolder(selItems As IList,
                                                 Pictures_Checkbox As Boolean,
                                                 SortFolderpath As String,
                                                 Save_MSG As Boolean,
                                                 Attchments As Boolean,
                                                 Remove_Flow_File As Boolean,
                                                 AppGlobals As IApplicationGlobals,
                                                 Optional StrRoot As String = "")
        Dim loc As String
        Dim FileSystem_LOC As String
        Dim FileSystem_LOC2 As String
        Dim FileSystem_DelLOC As String
        'Dim selItems            As Selection    ' A collection of Outlook item objects in a folder.
        'Dim selItems            As Collection
        Dim objItem As Object
        Dim MSG As MailItem
        Dim objFSO As Object       ' Computer's file system object.
        Dim objShell As Object       ' Windows Shell application object.
        Dim objFolder As Object       ' The selected folder object from Browse for Folder dialog box.
        Dim objSubFolders As Object
        Dim objNewFolder As Object
        Dim sortFolder As Outlook.Folder
        Dim folderCurrent As Outlook.Folder
        Dim strFolderPath As String = ""
        Dim i As Integer
        Dim oMailTmp As MailItem
        Dim strTemp As String
        Dim strAry() As String
        Dim strOutput(1) As String

        '******************
        '***INITIALIZE*****
        '******************
        Dim _globals As IApplicationGlobals = AppGlobals
        If (StrRoot = "") Then
            StrRoot = _globals.Ol.ArchiveRootPath
        End If
        'TODO: Eliminate following line once Path.Combine used below
        loc = StrRoot & "\"

        Dim _olApp As Outlook.Application = _globals.Ol.App
        Dim OlNS As Outlook.NameSpace = _globals.Ol.NamespaceMAPI


        If selItems.Count > 0 Then
            folderCurrent = GetCurrentExplorerFolder(_globals.Ol.App.ActiveExplorer(), selItems(0))
        Else
            folderCurrent = GetCurrentExplorerFolder(_globals.Ol.App.ActiveExplorer())
        End If
        If InStr(folderCurrent.FolderPath, _globals.Ol.Inbox.FolderPath) Then
            strFolderPath = _globals.FS.FldrFlow
        ElseIf InStr(folderCurrent.FolderPath, StrRoot) And folderCurrent.FolderPath <> StrRoot Then
            strFolderPath = folderCurrent.ToFsFolder(OlFolderRoot:=_globals.Ol.ArchiveRootPath, FsFolderRoot:=_globals.FS.FldrRoot)
            'strFolderPath = _globals.FS.FldrRoot & Right(folderCurrent.FolderPath, Len(folderCurrent.FolderPath) - Len(_globals.Ol.ArchiveRootPath) - 1)
        Else

        End If



        '*************************************************************************
        '************** SAVE ATTACHMENTS IF ENABLED*******************************
        '*************************************************************************
        Dim strTemp2 As String = ""
        'QUESTION: Original code allowed path to be an optional variable and then did something if a value was supplied that didn't match the archive root. Need to determine why and if new treatment loses functionality
        If StrRoot <> _globals.Ol.ArchiveRootPath Then
            strTemp2 = Right(_globals.Ol.ArchiveRootPath, Len(_globals.Ol.ArchiveRootPath) - Len(_globals.Ol.EmailRootPath) - 1)
            FileSystem_LOC = _globals.FS.FldrRoot & strTemp2 & "\" & SortFolderpath  'Parent Directory
        Else
            FileSystem_LOC = Path.Combine(_globals.FS.FldrRoot, SortFolderpath)
        End If

        FileSystem_DelLOC = _globals.FS.FldrRoot

        'If Save_PDF = True Then
        'Call SaveAsPDF.SaveMessageAsPDF(FileSystem_LOC, selItems)
        'End If

        If Save_MSG = True Then
            Call SaveMessageAsMSG(FileSystem_LOC, selItems)
        End If
        '



        '****Save Attachment to OneDrive directory****

        If Attchments = True Then
            'Email_SortSaveAttachment.SaveAttachmentsFromSelection(SavePath:=FileSystem_LOC, Verify_Action:=Pictures_Checkbox, selItems:=selItems, save_images:=Pictures_Checkbox, SaveMSG:=Save_MSG)
            SaveAttachmentsFromSelection(AppGlobals:=AppGlobals,
                                         SavePath:=FileSystem_LOC,
                                         Verify_Action:=Pictures_Checkbox,
                                         selItems:=selItems,
                                         save_images:=Pictures_Checkbox,
                                         SaveMSG:=Save_MSG)
        End If



        If Remove_Flow_File = True Then
            Call SaveAttachmentsFromSelection(AppGlobals:=AppGlobals, SavePath:=strFolderPath, DELFILE:=True, selItems:=selItems)
        End If



        '*************************************************************************
        '*********** LABEL EMAIL AS AUTOSORTED AND MOVE TO EMAIL FOLDER***********
        '*************************************************************************

        'If strTemp2 = "" Then Add_Recent(SortFolderpath)
        If strTemp2 = "" Then _globals.AF.RecentsList.AddRecent(SortFolderpath)
        loc = Path.Combine(StrRoot, SortFolderpath)
        sortFolder = New FolderHandler(_globals).GetFolder(loc) 'Call Function to turn text to Folder

        'Call Flag_Fields_Categories.SetCategory("Autosort")
        'Call Flag_Fields_Categories.CustomFieldID_Set("Autosort", "True")
        If sortFolder Is Nothing Then
            MsgBox(loc & " does not exist, skipping email move.")
        Else

            For i = selItems.Count - 1 To 0 Step -1
                If TypeOf selItems(i) Is Outlook.MailItem Then
                    If Not TypeOf selItems(i) Is Outlook.MeetingItem Then
                        MSG = selItems(i)
                        If strTemp2 = "" Then
                            'Email_AutoCategorize.UpdateForMove(MSG, SortFolderpath)
                            UpdateForMove(MSG, SortFolderpath, AppGlobals.AF.CTFList)
                        End If
                        On Error Resume Next
                        CustomFieldID_Set("Autosort", "True", SpecificItem:=MSG)
                        MSG.UnRead = False
                        MSG.Save()

                        oMailTmp = MSG.Move(sortFolder)

                        If Err.Number <> 0 Then
                            'TODO: ERROR LOGGING
                            'MsgBox("Error in " & SubNm & "-> MailItem.Move: " & Err.Number & " -> " & Err.Description & " ->" & Err.Source)
                            Err.Clear()
                        Else
                            If _globals.Ol.MovedMails_Stack Is Nothing Then _globals.Ol.MovedMails_Stack = New StackObjectVB
                            _globals.Ol.MovedMails_Stack.Push(MSG)
                            _globals.Ol.MovedMails_Stack.Push(oMailTmp)

                            'TODO: Change this into a JSON file
                            WriteCSV_StartNewFileIfDoesNotExist(_globals.FS.Filenames.EmailMoves, _globals.FS.FldrMyD)
                            strAry = CaptureEmailDetails(oMailTmp, _globals.Ol.ArchiveRootPath)
                            strOutput(1) = SanitizeArrayLineTSV(strAry)
                            FileIO2.Write_TextFile(_globals.FS.Filenames.EmailMoves, strOutput, _globals.FS.FldrMyD)
                        End If

                    End If
                End If

                'FireTimerReset
            Next i
        End If
    End Sub

    Private Function SanitizeArrayLineTSV(ByRef strOutput() As String) As String
        Dim strBuild As String = ""
        If strOutput.IsAllocated() Then
            Dim max As Integer = UBound(strOutput)
            For i = 1 To max
                Dim strTemp As String = strOutput(i)
                strTemp = Replace$(Trim$(strTemp), vbTab, "")
                strTemp = Replace$(strTemp, vbCrLf, " ")
                strTemp = Replace$(strTemp, vbLf, " ")

                strBuild = strBuild & vbTab & strTemp

            Next i
            If Len(strBuild) > 0 Then strBuild = Right(strBuild, Len(strBuild) - 1)
            Return strBuild
        Else
            Return ""
        End If
    End Function

    Private Sub WriteCSV_StartNewFileIfDoesNotExist(strFileName As String, strFileLocation As String)
        Dim strOutput() As String
        Dim strAryOutput(,) As String
        Dim objFSO As Object
        strOutput = Nothing
        If File.Exists(Path.Combine(strFileName, strFileLocation)) Then
            ReDim strAryOutput(13, 1)

            strAryOutput(1, 1) = "Triage"
            strAryOutput(2, 1) = "FolderName"
            strAryOutput(3, 1) = "Sent_On"
            strAryOutput(4, 1) = "From"
            strAryOutput(5, 1) = "To"
            strAryOutput(6, 1) = "CC"
            strAryOutput(7, 1) = "Subject"
            strAryOutput(8, 1) = "Body"
            strAryOutput(9, 1) = "fromDomain"
            strAryOutput(10, 1) = "Conversation_ID"
            strAryOutput(11, 1) = "EntryID"
            strAryOutput(12, 1) = "Attachments"
            strAryOutput(13, 1) = "FlaggedAsTask"

            Sanitize_Array(strAryOutput, strOutput)
            Write_TextFile(strFileName, strOutput, strFileLocation:=strFileLocation)

        End If
        Erase strOutput
        Erase strAryOutput
        objFSO = Nothing

    End Sub

    Private Sub Sanitize_Array(strAryOutput(,) As String, strOutput() As String)
        Dim i As Integer
        Dim j As Integer
        Dim maxi As Integer
        Dim maxj As Integer
        Dim strTemp As String

        If strAryOutput.IsAllocated() Then
            maxi = UBound(strAryOutput, 1)
            maxj = UBound(strAryOutput, 2)
            ReDim strOutput(maxj)

            For j = 1 To maxj
                For i = 1 To maxi
                    strTemp = strAryOutput(i, j)
                    strTemp = Replace$(Trim$(strTemp), vbTab, "")
                    strTemp = Replace$(strTemp, vbCrLf, " ")
                    strTemp = Replace$(strTemp, vbLf, " ")
                    strAryOutput(i, j) = strTemp
                    strOutput(j) = strOutput(j) & vbTab & strTemp
                Next i
                strOutput(j) = Right(strOutput(j), Len(strOutput(j)) - 1)
            Next j
        Else
            MsgBox("Empty Array in Sub Sanitize_Array")
        End If
    End Sub

    Private Sub UpdateForMove(MSG As MailItem, fldr As String, CTFList As CtfIncidenceList)
        Dim Inc_Num As Integer
        Dim i, j As Integer
        Dim tmp_CTF_Map As Conversation_To_Folder
        Dim tmpCCT, tmpFDR As String
        Dim updated As Boolean


        updated = False
        Inc_Num = CTFList.CTF_Incidence_FIND(MSG.ConversationID)                        'Check to see if the conversation id is already in the incidence matrix

        If Inc_Num = 0 Then                                                     'If it is not in the matrix,
            CTFList.CTF_Inc_Ct += 1                                         'increase matrix record count
            ReDim Preserve CTFList.CTF_Inc(CTFList.CTF_Inc_Ct)                                  'and expand the matrix

            tmp_CTF_Map.Email_Conversation_Count = 1
            tmp_CTF_Map.Email_Conversation_ID = MSG.ConversationID
            tmp_CTF_Map.Email_Folder = fldr

            Call CTFList.CTF_Incidence_INIT(CTFList.CTF_Inc_Ct)                                 'Initialize Variable
            Call CTFList.CTF_Incidence_SET(CTFList.CTF_Inc_Ct, 1, 1, tmp_CTF_Map)                'Map Variable in top position

        Else

            With CTFList.CTF_Inc(Inc_Num)

                For i = 1 To .Folder_Count
                    If .Email_Folder(i) = fldr Then
                        .Email_Conversation_Count(i) = 1 + .Email_Conversation_Count(i)
                        updated = True
                        If i > 1 Then
                            For j = i To 2 Step -1
                                If .Email_Conversation_Count(j) > .Email_Conversation_Count(j - 1) Then
                                    tmpCCT = .Email_Conversation_Count(j)
                                    tmpFDR = .Email_Folder(j)
                                    .Email_Conversation_Count(j) = .Email_Conversation_Count(j - 1)
                                    .Email_Folder(j) = .Email_Folder(j - 1)
                                    .Email_Conversation_Count(j - 1) = tmpCCT
                                    .Email_Folder(j - 1) = tmpFDR
                                Else
                                    Exit For
                                End If
                            Next j
                        End If
                        If updated = True Then Exit For
                    End If
                Next i

            End With

            If updated = False Then

                With tmp_CTF_Map
                    .Email_Conversation_Count = 1
                    .Email_Conversation_ID = MSG.ConversationID
                    .Email_Folder = fldr
                End With

                Call CTFList.CTF_Inc_Position_ADD(Inc_Num, tmp_CTF_Map)                     'If it is in the matrix, add it in the right slot

            End If

        End If

        Call Subject_Map_Add(MSG.Subject, fldr)
    End Sub

    'Private Sub Add_Recent(sortFolder As String)
    '    Throw New NotImplementedException()
    'End Sub

    'Private Sub SaveAttachmentsFromSelection(strFolderPath As String, v As Boolean, Optional value As Object = Nothing, Optional selItems As IList = Nothing)
    '    Throw New NotImplementedException()
    'End Sub

    'Private Sub SaveAttachmentsFromSelection(SavePath As String, Verify_Action As Boolean, selItems As IList, save_images As Boolean, SaveMSG As Boolean)
    '    Throw New NotImplementedException()
    'End Sub

    Private Sub SaveMessageAsMSG(fileSystem_LOC As String, selItems As IList)
        Throw New NotImplementedException()
    End Sub

    Private Function GetCurrentExplorerFolder(ActiveExplorer As Outlook.Explorer, Optional objItem As Object = Nothing) As Folder
        If objItem Is Nothing Then
            objItem = ActiveExplorer.Selection.Item(0)
        End If

        If TypeOf objItem Is MailItem Then
            Dim OlMail As MailItem = objItem
            Return OlMail.Parent

        ElseIf TypeOf objItem Is AppointmentItem Then
            Dim OlAppointment As Outlook.AppointmentItem = objItem
            Return OlAppointment.Parent

        ElseIf TypeOf objItem Is MeetingItem Then
            Dim OlMeeting As Outlook.MeetingItem = objItem
            Return OlMeeting.Parent

        ElseIf TypeOf objItem Is TaskItem Then
            Dim OlTask As TaskItem = objItem
            Return OlTask.Parent

        Else
            Return Nothing
        End If

    End Function

    Public Sub Cleanup_Files()
        Throw New NotImplementedException
    End Sub

    'Public Function DialogueThrowNotImplemented() As Boolean
    '    Return MsgBox("")
    'End Function

End Module

File: ToDoModel/Email Utilities/SortItemsToExistingFolder.cs
using System;
using System.Collections;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using System.Windows.Forms;
using Microsoft.Office.Interop.Outlook;
using UtilitiesCS;
using UtilitiesCS.OutlookExtensions;
using System.Collections.Generic;

namespace ToDoModel
{

    public static class SortItemsToExistingFolder
    {
        public static void InitializeSortToExisting(string InitType = "Sort", bool QuickLoad = false, bool WholeConversation = true, string strSeed = "", object objItem = null)
        {
            throw new NotImplementedException();
        }

        public static void Run(IList<MailItem> selItems, bool picturesCheckbox, string sortFolderpath, bool saveMsg, bool attchments, bool removeFlowFile, IApplicationGlobals appGlobals, string strRoot = "")
        {
            string loc;
            string FileSystem_LOC;
            string FileSystem_LOC2;
            string FileSystem_DelLOC;
            // Dim selItems            As Selection    ' A collection of Outlook item objects in a folder.
            // Dim selItems            As Collection
            object objItem;
            MailItem MSG;
            object objFSO;       // Computer's file system object.
            object objShell;       // Windows Shell application object.
            object objFolder;       // The selected folder object from Browse for Folder dialog box.
            object objSubFolders;
            object objNewFolder;
            Folder sortFolder;
            Folder folderCurrent;
            string strFolderPath = "";
            int i;
            MailItem oMailTmp;
            string strTemp;
            string[] strAry;
            var strOutput = new string[2];

            // ******************
            // ***INITIALIZE*****
            // ******************
            var _globals = appGlobals;
            if (string.IsNullOrEmpty(strRoot))
            {
                strRoot = _globals.Ol.ArchiveRootPath;
            }
            // TODO: Eliminate following line once Path.Combine used below
            loc = strRoot + @"\";

            var _olApp = _globals.Ol.App;
            var OlNS = _globals.Ol.NamespaceMAPI;


            if (selItems.Count > 0)
            {
                folderCurrent = GetCurrentExplorerFolder(_globals.Ol.App.ActiveExplorer(), selItems[0]);
            }
            else
            {
                folderCurrent = GetCurrentExplorerFolder(_globals.Ol.App.ActiveExplorer());
            }
            if (folderCurrent.FolderPath.Contains(_globals.Ol.Inbox.FolderPath))
            {
                strFolderPath = _globals.FS.FldrFlow;
            }
            else if (folderCurrent.FolderPath.Contains(strRoot) & (folderCurrent.FolderPath != strRoot ))
            {
                strFolderPath = folderCurrent.ToFsFolder(OlFolderRoot: _globals.Ol.ArchiveRootPath, FsFolderRoot: _globals.FS.FldrRoot);
            }
            // strFolderPath = _globals.FS.FldrRoot & Right(folderCurrent.FolderPath, Len(folderCurrent.FolderPath) - Len(_globals.Ol.ArchiveRootPath) - 1)
            else
            {

            }



            // *************************************************************************
            // ************** SAVE ATTACHMENTS IF ENABLED*******************************
            // *************************************************************************
            string strTemp2 = "";
            // QUESTION: Original code allowed path to be an optional variable and then did something if a value was supplied that didn't match the archive root. Need to determine why and if new treatment loses functionality
            if ((strRoot ?? "") != (_globals.Ol.ArchiveRootPath ?? ""))
            {
                strTemp2 = _globals.Ol.ArchiveRootPath.Substring(_globals.Ol.EmailRootPath.Length);
                FileSystem_LOC = _globals.FS.FldrRoot + strTemp2 + @"\" + sortFolderpath;  // Parent Directory
            }
            else
            {
                FileSystem_LOC = Path.Combine(_globals.FS.FldrRoot, sortFolderpath);
            }

            FileSystem_DelLOC = _globals.FS.FldrRoot;

            // If Save_PDF = True Then
            // Call SaveAsPDF.SaveMessageAsPDF(FileSystem_LOC, selItems)
            // End If

            if (saveMsg == true)
            {
                SaveMessageAsMSG(FileSystem_LOC, selItems);
            }
            // 



            // ****Save Attachment to OneDrive directory****

            if (attchments == true)
            {
                // Email_SortSaveAttachment.SaveAttachmentsFromSelection(SavePath:=FileSystem_LOC, Verify_Action:=Pictures_Checkbox, selItems:=selItems, save_images:=Pictures_Checkbox, SaveMSG:=Save_MSG)
                SaveAttachmentsModule.SaveAttachmentsFromSelection(AppGlobals: appGlobals, SavePath: FileSystem_LOC, Verify_Action: picturesCheckbox, selItems: selItems, save_images: picturesCheckbox, SaveMSG: saveMsg);
            }



            if (removeFlowFile == true)
            {
                SaveAttachmentsModule.SaveAttachmentsFromSelection(AppGlobals: appGlobals, SavePath: strFolderPath, DELFILE: true, selItems: selItems);
            }



            // *************************************************************************
            // *********** LABEL EMAIL AS AUTOSORTED AND MOVE TO EMAIL FOLDER***********
            // *************************************************************************

            // If strTemp2 = "" Then Add_Recent(SortFolderpath)
            if (string.IsNullOrEmpty(strTemp2))
                _globals.AF.RecentsList.AddRecent(sortFolderpath);
            loc = Path.Combine(strRoot, sortFolderpath);
            sortFolder = new FolderHandler(_globals).GetFolder(loc); // Call Function to turn text to Folder

            // Call Flag_Fields_Categories.SetCategory("Autosort")
            // Call Flag_Fields_Categories.SetUdf("Autosort", "True")
            if (sortFolder is null)
            {
                MessageBox.Show(loc + " does not exist, skipping email move.");
            }
            else
            {

                for (i = selItems.Count - 1; i >= 0; i -= 1)
                {
                    if (selItems[i] is MailItem)
                    {
                        if (!(selItems[i] is MeetingItem))
                        {
                            MSG = (MailItem)selItems[i];
                            if (string.IsNullOrEmpty(strTemp2))
                            {
                                // Email_AutoCategorize.UpdateForMove(MSG, SortFolderpath)
                                UpdateForMove(MSG, sortFolderpath, appGlobals.AF.CtfMap, appGlobals.AF.SubjectMap);
                            };
                            try
                            {
                                MSG.SetUdf("Autosort", "True");
                                MSG.UnRead = false;
                                MSG.Save();

                                oMailTmp = (MailItem)MSG.Move(sortFolder);
                                CaptureMoveDetails(MSG, oMailTmp, strOutput, _globals);
                            }
                            catch (System.Exception e)
                            {
                                Debug.WriteLine(e.Message);
                                Debug.WriteLine(e.StackTrace);
                            }
                        }
                    }
                }
            }
        }

        private static void CaptureMoveDetails(MailItem MSG, MailItem oMailTmp, string[] strOutput, IApplicationGlobals _globals)
        {
            if (_globals.Ol.MovedMails_Stack is null)
                _globals.Ol.MovedMails_Stack = new StackObjectCS<object>();
            _globals.Ol.MovedMails_Stack.Push(MSG);
            _globals.Ol.MovedMails_Stack.Push(oMailTmp);

            // TODO: Change this into a JSON file
            WriteCSV_StartNewFileIfDoesNotExist(_globals.FS.Filenames.EmailMoves, _globals.FS.FldrMyD);
            //string[] strAry = CaptureEmailDetailsModule.CaptureEmailDetails(oMailTmp, _globals.Ol.ArchiveRootPath);
            string[] strAry = oMailTmp.Details(_globals.Ol.ArchiveRootPath);
            strOutput[1] = SanitizeArrayLineTSV(ref strAry);
            FileIO2.WriteTextFile(_globals.FS.Filenames.EmailMoves, strOutput, _globals.FS.FldrMyD);
        }

        //private static string SanitizeArrayLineTSV(ref string[] strOutput)
        //{
        //    string strBuild = "";
        //    if (strOutput.IsInitialized())
        //    {
        //        int max = strOutput.Length;
        //        for (int i = 1, loopTo = max; i <= loopTo; i++)
        //        {
        //            string strTemp = StripTabsCrLf(strOutput[i]);
        //            strBuild = strBuild + "\t" + strTemp;

        //        }
        //        if (strBuild.Length > 0)
        //            strBuild = strBuild.Substring(1);
        //        return strBuild;
        //    }
        //    else
        //    {
        //        return "";
        //    }
        //}

        private static string SanitizeArrayLineTSV(ref string[] strOutput)
        {
            if (strOutput.IsInitialized())
            {
                return string.Join("\t",strOutput
                             .Where(s => !string.IsNullOrEmpty(s))
                             .Select(s => StripTabsCrLf(s))
                             .ToArray());
            }
            else { return ""; }
        }

        internal static string StripTabsCrLf(string str)
        {
            var _regex = new Regex(@"[\t\n\r]*");
            string result = _regex.Replace(str, " ");

            // ensure max of one space per word
            _regex = new Regex(@"  +");
            result = _regex.Replace(result, " ");
            result = result.Trim();
            return result;
        }

        private static void WriteCSV_StartNewFileIfDoesNotExist(string strFileName, string strFileLocation)
        {
            string[] strOutput = null;
            string[,] strAryOutput;
            if (File.Exists(Path.Combine(strFileName, strFileLocation)))
            {
                strAryOutput = new string[14, 2];

                strAryOutput[1, 1] = "Triage";
                strAryOutput[2, 1] = "FolderName";
                strAryOutput[3, 1] = "Sent_On";
                strAryOutput[4, 1] = "From";
                strAryOutput[5, 1] = "To";
                strAryOutput[6, 1] = "CC";
                strAryOutput[7, 1] = "Subject";
                strAryOutput[8, 1] = "Body";
                strAryOutput[9, 1] = "fromDomain";
                strAryOutput[10, 1] = "Conversation_ID";
                strAryOutput[11, 1] = "EntryID";
                strAryOutput[12, 1] = "Attachments";
                strAryOutput[13, 1] = "FlaggedAsTask";

                SanitizeArray(strAryOutput, ref strOutput);
                FileIO2.WriteTextFile(strFileName, strOutput, folderpath: strFileLocation);

            }
            strOutput = null;
            strAryOutput = null;
        }

        private static void SanitizeArray(string[,] strAryOutput, ref string[] strOutput)
        {
            if (strAryOutput == null) 
            {
                Debug.WriteLine($"The array {nameof(strAryOutput)} is empty.");
            }
            else
            {
                for (int j = 0; j < strAryOutput.GetLength(0); j++)
                {
                    strOutput[j] = string.Join("\t", strAryOutput
                                         .SliceRow(j)
                                         .Where(s => !string.IsNullOrEmpty(s))
                                         .Select(s => StripTabsCrLf(s))
                                         .ToArray());
                }
            }
        }

        private static void UpdateForMove(MailItem mailItem, string fldr, CtfMap ctfMap, ISubjectMapSL subMap)
        {
            ctfMap.Add(mailItem.ConversationID, fldr, 1);
            subMap.Add(mailItem.Subject, fldr);
        }
                
        private static void SaveMessageAsMSG(string fileSystem_LOC, IList<MailItem> selItems)
        {
            throw new NotImplementedException();
        }

        private static Folder GetCurrentExplorerFolder(Explorer ActiveExplorer, object objItem = null)
        {
            if (objItem is null)
            {
                objItem = ActiveExplorer.Selection[0];
            }

            if (objItem is MailItem)
            {
                MailItem OlMail = (MailItem)objItem;
                return (Folder)OlMail.Parent;
            }

            else if (objItem is AppointmentItem)
            {
                AppointmentItem OlAppointment = (AppointmentItem)objItem;
                return (Folder)OlAppointment.Parent;
            }

            else if (objItem is MeetingItem)
            {
                MeetingItem OlMeeting = (MeetingItem)objItem;
                return (Folder)OlMeeting.Parent;
            }

            else if (objItem is TaskItem)
            {
                TaskItem OlTask = (TaskItem)objItem;
                return (Folder)OlTask.Parent;
            }

            else
            {
                return null;
            }

        }

        public static void Cleanup_Files()
        {
            // Call WRITE_Text_File     - Writes to the recents list
            // Call Email_AutoCategorize.CTF_Incidence_Text_File_WRITE - Writes to the CTF_Incidence file   
            // Call Email_AutoCategorize.Subject_MAP_Text_File_WRITE - Writes to the Subject_MAP file
        }

        // Public Function DialogueThrowNotImplemented() As Boolean
        // Return MsgBox("")
        // End Function

    }
}